File: project files/server/src/models/Product.js
const mongoose = require('mongoose');

const variantSchema = new mongoose.Schema(
  {
    variantId: { type: mongoose.Schema.Types.ObjectId, auto: true },
    weight: { type: String, required: true, trim: true },
    price: { type: Number, required: true, min: 0 },
    MRP: { type: Number, required: true, min: 0 },
    stock: { type: Number, required: true, min: 0, default: 0 },
    skuSuffix: { type: String, required: true, trim: true }
  },
  { _id: false }
);

const productSchema = new mongoose.Schema(
  {
    name: { type: String, required: true, trim: true },
    slug: { type: String, required: true, unique: true, index: true, trim: true },
    SKU: { type: String, required: true, unique: true, trim: true },
    brand: { type: String, required: true, index: true, trim: true },
    categoryId: { type: mongoose.Schema.Types.ObjectId, ref: 'Category', required: true, index: true },
    subCategoryId: { type: mongoose.Schema.Types.ObjectId, ref: 'Category', required: true, index: true },
    sellerId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true, index: true },
    variants: { type: [variantSchema], validate: [(arr) => arr.length > 0, 'At least one variant is required'] },
    rating: { type: Number, default: 0, min: 0, max: 5 },
    totalReviews: { type: Number, default: 0, min: 0 },
    salesCount: { type: Number, default: 0, min: 0, index: true },
    images: [{ type: String, trim: true }],
    description: { type: String, default: '' },
    tags: [{ type: String, trim: true }],
    isFeatured: { type: Boolean, default: false, index: true },
    isActive: { type: Boolean, default: true, index: true },
    adminApproved: { type: Boolean, default: false },
    adminReviewNote: { type: String, default: '' },
    lowStockThreshold: { type: Number, default: 10, min: 0 }
  },
  { timestamps: true, toJSON: { virtuals: true }, toObject: { virtuals: true } }
);

productSchema.virtual('discountPercent').get(function discountPercent() {
  let maxDiscount = 0;
  for (const variant of this.variants || []) {
    if (variant.MRP > 0 && variant.MRP >= variant.price) {
      const pct = Math.round(((variant.MRP - variant.price) / variant.MRP) * 100);
      if (pct > maxDiscount) {
        maxDiscount = pct;
      }
    }
  }
  return maxDiscount;
});

productSchema.index({ name: 'text', description: 'text' });
productSchema.index({ categoryId: 1 });
productSchema.index({ subCategoryId: 1 });
productSchema.index({ brand: 1 });
productSchema.index({ isFeatured: 1 });
productSchema.index({ isActive: 1 });
productSchema.index({ salesCount: 1 });
productSchema.index({ sellerId: 1 });

module.exports = mongoose.model('Product', productSchema);

File: spring_back/src/main/java/com/groceryapp/model/Product.java
package com.groceryapp.model;

import jakarta.persistence.CollectionTable;
import jakarta.persistence.Column;
import jakarta.persistence.ElementCollection;
import jakarta.persistence.Embeddable;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.Table;
import jakarta.validation.constraints.NotBlank;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "products")
public class Product {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private String id;

    @NotBlank
    private String name;

    @Column(unique = true)
    private String slug;

    @Column(name = "SKU", unique = true)
    private String SKU;

    private String brand;

    @Column(name = "category_id")
    private String categoryId;

    @Column(name = "sub_category_id")
    private String subCategoryId;

    @Column(name = "seller_id")
    private String sellerId;

    @ElementCollection
    @CollectionTable(name = "product_variants", joinColumns = @JoinColumn(name = "product_id"))
    @Builder.Default
    private List<Variant> variants = new ArrayList<>();

    private Double rating;
    private Integer totalReviews;

    @Column(name = "sales_count")
    private Integer salesCount;

    @ElementCollection
    @CollectionTable(name = "product_images", joinColumns = @JoinColumn(name = "product_id"))
    @Builder.Default
    private List<String> images = new ArrayList<>();

    private String description;

    @ElementCollection
    @CollectionTable(name = "product_tags", joinColumns = @JoinColumn(name = "product_id"))
    @Builder.Default
    private List<String> tags = new ArrayList<>();

    @Column(name = "is_featured")
    private Boolean isFeatured;

    @Column(name = "is_active")
    private Boolean isActive;

    private Boolean adminApproved;
    private String adminReviewNote;
    private Integer lowStockThreshold;
    private Instant createdAt;
    private Instant updatedAt;

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    @Embeddable
    public static class Variant {
        private String variantId;
        private String weight;
        private Double price;
        @Column(name = "MRP")
        private Double MRP;
        private Integer stock;
        private String skuSuffix;
    }
}
